File: backend/model/order.rb
class Order
  attr_reader :customer, :items, :payment, :address, :closed_at

  def initialize(customer, overrides = {})
    @customer = customer
    @items = []
    @order_item_class = overrides.fetch(:item_class) { OrderItem }
    @address = overrides.fetch(:address) { Address.new(zipcode: '45678-979') }
    @discount = 0
  end

  def add_product(product)
    @items << @order_item_class.new(order: self, product: product)
  end

  def total_amount
    @items.map(&:total).inject(:+) - @discount
  end

  def close(closed_at = Time.now)
    @closed_at = closed_at
    @items.each do |item|
      item.product.actions.after_payment(item)
    end
  end

  def prepare_payment
    @items.each do |item|
      item.product.actions.before_payment(item)
    end
  end

  def add_discount(discount)
    @discount += discount
  end
end

File: backend/src/main/kotlin/com/creditas/model/Order.kt
package com.creditas.model

import com.creditas.service.fulfillment.FulfillmentStrategy
import java.util.*

class Order(val customer: Customer, val address: Address) {
    val items = mutableListOf<OrderItem>()
    var closedAt: Date? = null
        private set
    var payment: Payment? = null
        private set
    val totalAmount
        get() = items.sumByDouble { it.total }
    val totalItems
        get() = items.sumBy { it.quantity }

    fun addProduct(product: Product, quantity: Int) {
        if (quantity <= 0)
            throw IllegalArgumentException("Quantity must be greater than 0!")
        var productAlreadyAdded = items.any { it.product == product }
        if (productAlreadyAdded)
            throw Exception("The product have already been added. Change the amount if you want more.")

        items.add(OrderItem(product, quantity))
    }

    fun pay(method: PaymentMethod) {
        if (payment != null)
            throw Exception("The order has already been paid!")

        if (items.count() == 0)
            throw Exception("Empty order can not be paid!")

        payment = Payment(this, method)

        close()
    }

    private fun close() {
        closedAt = Date()
    }
}
